demux/demux_lavf.c

```diff
@@ -736,6 +769,72 @@ static void handle_new_stream(demuxer_t *demuxer, int i)
     select_tracks(demuxer, i);
 }
 
+static void detect_program_changes(demuxer_t *demuxer)
+{
+    lavf_priv_t *priv = demuxer->priv;
+    struct sh_stream *old_stream = NULL, *new_stream = NULL;
+    AVProgram *program;
+    int i, j;
+
+    if (priv->avfc->nb_programs < 1)
+        return;
+    if (priv->program.num == 0)
+        return;
+
+    for (i = 0; i < priv->avfc->nb_programs; i++)
+        if (priv->avfc->programs[i]->id == priv->program.num)
+            break;
+    if (i == priv->avfc->nb_programs)
+        return;
+    program = priv->avfc->programs[i];
+
+    // skip unless program changed
+    if (program->pcr_pid == priv->program.pcr_pid &&
+        program->nb_stream_indexes == priv->program.nb_stream_indexes) {
+        return;
+    }
+
+    MP_VERBOSE(demuxer, "Detected program change for %d: "
+                        "(pcr=%d,nb=%d) -> (pcr=%d,nb=%d) \n",
+               program->program_num,
+               priv->program.pcr_pid, priv->program.nb_stream_indexes,
+               program->pcr_pid, program->nb_stream_indexes);
+
+    priv->program = (lavf_program_t){
+        .num = program->program_num,
+        .pcr_pid = program->pcr_pid,
+        .nb_stream_indexes = program->nb_stream_indexes,
+    };
+
+    for (i = 0; i < priv->num_streams; i++) {
+        struct sh_stream *stream = priv->streams[i];
+        if (!stream || stream->type != STREAM_AUDIO)
+            continue;
+
+        int found = 0;
+        for (j = 0; j < program->nb_stream_indexes; j++) {
+            if (program->stream_index[j] == stream->ff_index) {
+                found = 1;
+                break;
+            }
+        }
+
+        if (found) {
+            if (!new_stream) new_stream = stream;
```

Comment:
Instead of just using the first audio track found in the updated program, we could make this smarter to try to pick the track with the same language/channels, or the same offset.

IMHO this is fine for a first pass as-is, since worst case the audio changes to something you didn't expect and you have to manually switch the audio track to something else.